flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java

```diff
@@ -112,7 +111,7 @@ public void accept(Throwable t) {
 	public abstract void addSplits(List<SplitT> splitsToAdd);
 
 	protected void startFetcher(SplitFetcher<E, SplitT> fetcher) {
-		executors.submit(new ThrowableCatchingRunnable(errorHandler, fetcher));
+		executors.submit(fetcher);
```

Comment:
I am curious. Is there a reason that we replace the `ThrowableCatchingRunnable` wrapper with the explicit error handler passed to the split fetcher here?

Reply:
We need error-setting(`errorHandler.accept(t)`) happens-before fetcher-removing(`shutdownHook.run()`) so that newly added error-checking wouldn't lost it. In order to achieve this, we need move `errorHandler` to `SplitFetcher` where `shutdownHook` locates in. After movement, `ThrowableCatchingRunnable` is useless.

Besides above semantic changes, I think it would be good to drop extra layer between two concurrent components `SourceReaderBase` and `SplitFetcher`.

Reply:
I personally like the change - less wrapping tends to make code navigation easier. But this is a matter of taste here, either way looks totally fine.

Reply:
But I think @kezhuw has a point - having this in the same class (rather than in a separat class `ThrowableCatchingRunnable`) is a good if there is a strict contract between the exception handling and closing. Otherwise the exception handling is factored out into a separate class (`ThrowableCatchingRunnable`) but with an implicit contract on how it must be used, to guarantee the happens-before relationship.

Reply:
@becketqin I see new code in `SplitFetcher.run` throwing exception after `shutdownHook.run`. Is there any specific reason where run `splitReader.close` after shutdownHook ? Or, in another word, does exception from `splitReader.close` should fail job ? I will do rebase to solve conflict, please take a took after rebasing.

Reply:
@kezhuw Thanks for the explanation. Good point.